Custom Get Item Messages for Randomizer Custom Items

leggettc18 opened the pull request:
This is my in-progress addition of custom message handling for custom items that we may introduce later on, such as the magic bean pack, or the keysanity keys down the line. Basically, an unused Get Item TextId (0xF8) is used to escape from the normal message flow in z_message_PAL.c. This ends up calling out to a function I have defined in the new `randomizer_custom_messages.cpp` (and corresponding `.h`) file. This file contains a few helper functions for certain special characters (such as for changing the colors or adding newlines), as well as the function itself that returns custom text for specific GetItemIDs. That function just switch-cases on the GetItemID, and for each case of that it switch-cases on the language. There may be a better way to do this, this is just the first go-round.

I have a video demonstrating this with the magic bean pack custom item (which I have on another branch and have been merging this branch in periodically), but GitHub wasn't letting me upload it for some reason. I'll try again in a comment.

MelonSpeedruns committed:
Free Camera (#337)

* wip free cam

* Almost done, needs collision still

* Added free cam behind cvar

* added WIP collision

* Fixed & implemented "Manual mode" from WW & TP

* Fixed camera not rotating when Link is moving

* fixed initialized camera rotation

* Fixed camera getting stuck + made it smoother

* reduced deadzone

* fixed epona camera height + added WW z-target free camera

* Adjusted player camera height & fixed fov

* Fixed camera roll

* fixed fov when moving the camera while in z-target

* Camera resets to Auto when going through doors or changing maps

* Fixed building

* touch

* more touch work

* Added WIP mouse support to the free cam

* gui stuff

* fixed building

* fixed building error

* ok fixed building for real this time

* oops

* Fix compilation issues

* removed mouse stuff that magically appeared in this branch

* smoothed out stick values & removed remains of mouse support

* re-added manual camera when pressing Z

* reduced minimum Y position of camera

* Addressed dcsv's nitpicks

* part 2

* oops

Co-authored-by: David Chavez <david@dcvz.io>

leggettc18 committed:
Rando: Allows Malon's Item Check to be obtained by pulling out the Ocarina. [FIXED PR] (#672)

* Fixes using the Ocarina to get the check from Malon.

Still some cleanup to do here. For some reason the player can shield before receiving the check. It doesn't set the flag if the player does that so they can still try again, but would prefer a different solution if possible.

* Prevents Shielding from blocking the Item_Give from happening.

* Code Cleanup and comments explaining the new rando flow.

* Removes inventory check when pulling out Ocarina

This allows OI to properly give the check, which is important for Glitched logic later down the line. Talking to Malon still requires the Ocarina in your inventory.

* Prevents non-malon textboxes from triggering the check.

Also adds a comment explaining the condtional for getting the check from talking to Malon since it got pretty long.

* Actually fixes checking for text boxes.

* Relocates a comment for improved clarity.

aMannus committed:
Fix Rando Water Temple Softlock (#665)

We use 3DS logic to generate item placement, but didn't have this specific door in Water Temple unlocked from the beginning like 3DS does.

This meant that if people took specific paths through the temple, they could softlock themselves by missing a key.

aMannus committed:
Rando: GtG and carpenter prompts skip (#663)

* Skip gtg and carpenter prompt

For rando. Tested and just works.

* Fixed missing break

vaguerant committed:
Hide debug overlay behind gDebugEnabled (#660)

dcvz committed:
Introduce App Directory Path (#572)

* Introduce app directory path concept

* macos: Remove hacky way of using applicaiton directory

* Update the new SaveManager

* Address stack user after return

* Remove unecessary property

* Use std::string for filepath

* Improve clang specific detections

* Use new path system for imgui files

* Improve helper for getting relative paths

Baoulettes committed:
fix hidden wnd (#744)

1SDANi committed:
Split damage multiplication into its own PR (#656)

* Split damage multiplication into its own PR

* Found a more elegant implementation of the powers char*[]

ChristopherJTrent committed:
Update Keese labels

Baoulettes committed:
.xiF slebaL

leggettc18 committed:
Fixes Maps, Compasses, and Boss Keys in Vanilla. (#751)

leggettc18 committed:
Early version of custom messages.

leggettc18 committed:
Makes sure custom get item textboxes match.

Specifically, they are all bottom of the screen and blue.

NEstelami committed:
Fixed soundfont issues

aMannus committed:
Skip warp song cutscenes in rando (#664)

Does it by skipping to the last part of the cutscene data. Tested on all songs, both adult and child.

briaguya0 committed:
don't spawn blocking mido after we've already shown him the sword/shield (#675)

Co-authored-by: briaguya <briaguya@alice>

KiritoDv committed:
Controller Configuration UI and JSON Config (#760)

* Initial controller hud ui

* Reverted fbdemo changes

* Moved config to json and implemented controller config

* fix build on linux, gitignore new config file

* fix build

* Fix compilation and file directory paths

* Call save on cvar save

* Fixed cvar loading and added deck slots to the config

* Changed control deck port 0 to use a physical device by default

* Added gyro and rumble & fixed loading errors

* Save config on toggle menubar

* fix linux build

* Fixed drift calculation

* Controller config now saves when pressing F1

* Removed ExitGame hook from ImGuiImpl

* Moved mappings to a map

* Added GetKeyName

* untranslate scancodes

* Fixed hud layout on keyboard device

* Fixed keyboard read on hud

* Fixed crash when reloading controllers

* Removed ConfigFile and changed file extension

* Changed Dummy to Disconnected and fixed filters

* Removed function leftover

* Changed ControllerHud to InputEditor

Co-authored-by: briaguya <briaguya@alice>
Co-authored-by: David Chavez <david@dcvz.io>

leggettc18 committed:
Refactors custom message code to its own file

Kenix3 committed:
Enough! My ship sails in the morning.

leggettc18 committed:
Renames the separated custom messages files to prevent conflict.

leggettc18 committed:
Merge branch 'develop' of https://github.com/HarbourMasters/Shipwright into rando-custom-messages

KiritoDv committed:
Fixed menubar items position (#763)

* Fixed menubar items position

* Reverted tooltip tab position

leggettc18 committed:
Merge branch 'develop' of https://github.com/HarbourMasters/Shipwright into rando-custom-messages

agamache committed:
Fixes macOS randomizer functionality with App Directory (#761)

* Fixes macOS randomizer functionality with App Directory

* Fix windows build

* Update soh/soh/Enhancements/randomizer/3drando/rando_main.cpp

* Update soh/soh/Enhancements/randomizer/3drando/spoiler_log.cpp

* Revert band-aid fix

Co-authored-by: Kenix3 <kenixwhisperwind@gmail.com>

leggettc18 committed:
Fixes newline function to return correct character.

dcvz committed:
Fix migration cvar deletion path (#765)

Kenix3 committed:
Merge branch 'develop' of https://github.com/HarbourMasters/Shipwright into develop

leggettc18 commented:
https://user-images.githubusercontent.com/21345721/178908845-5311952b-90a0-41d5-a3d5-3a0ecc7dd518.mp4

Demo of custom text with the Magic Bean Pack Item.

Sirius902 committed:
Various controller fixes (#771)

* Fix controller

* Also fix rumble strength being a bool

* Remove ControllerHud.cpp

* Downgrade platform toolset back to previous version

* Fix gyro

* Fix bug that makes binding axes difficult and clear buttons before reading

* Exaggerate gyro display and adjust stick binding threshold

* Initialize drift thresholds

leggettc18 committed:
Merge branch 'develop' of https://github.com/HarbourMasters/Shipwright into rando-custom-messages

leggettc18 commented:
For reference, this is the randomizer_custom_messages.cpp from my magic bean pack repo (seen in the demo video) that is currently utilizing this system. I had some cleanup to do before I pushed it to GitHub.

https://github.com/leggettc18/Shipwright/blob/rando-bean-pack/soh/soh/Enhancements/randomizer/randomizer_custom_messages.cpp

leggettc18 committed:
Added an extra helper function and removed instant text control code from ITEM_OBTAINED

briaguya0 commented:
> randomizer_custom_messages.cpp

i like this! it might be worth looking into trying to pull the hint logic out of `z_message_pal` and into there too

leggettc18 commented:
Agreed, any other comments about organization are welcome. Currently I just kind of put them in the Randomizer class, but I'm not sure if you'd want them to be separate or what. I borrowed a lot of the helper functions from 3ds rando code and adapted it (3DS control codes seem to be different). I do see that your code for formatting Hint Text has a lot of useful stuff as well that I might pull out, like the characters with diacritics that I was having issues with.

I'll move your functions for hint text there so we can see how that looks, and probably will write a function for Skulltula tokens as well since those already have a manually intercepted textId. That part might need to be revisited with Skullsanity though, not sure if getting the tokens out of a chest will behave any differently in terms of intercepting the text.

briaguya0 commented:
> probably will write a function for Skulltula tokens as well since those already have a manually intercepted textId

hmm, since the skulls aren't a rando specific thing it's making me wonder if having custom messages live in the rando directory makes sense, or if we should put it somewhere else (no ideas for where off the top of my head rn though)

aMannus commented:
Well, this is exciting.

leggettc18 commented:
Hmmm good point, rando is where we'll use it the most but I forgot that the Skulltula text works outside of rando as well. More General custom text could also be useful for modders since unless I'm missing something, adding/patching text into main message table in the OTR is... not very accessible.

aMannus commented:
Yeah i think having a seperate place for custom text that's usable elsewhere and then use that wherever we need it on the rando side is the play here. I doubt the edited skull token text is going to be the last feature that will be able to use this even outside of rando.

And yeah directly editing the message table somehow from the OTR sounds like hell.

vaguerant commented:
I just realized that the rando merge has made #471 somewhat redundant as that also would have added Skulltula token count message.

The PR version does offer one addition over what rando (and now dev) has, in that the message box is instant and auto-closes like in N64 rando if you have the no-freeze enhancement enabled.

Is there anything salvageable here, or is the implementation too different to what's already in dev now?

leggettc18 commented:
Hmm, I'm not sure whether this PR will end up including Skulltula text or not. I'm not sure yet whether the scope of this is just rando get item text, or a more general custom messages system. If it's just rando stuff then the Skulltula text will likely stay where it is and the auto-dismissal of the text box can be implemented in place, if this does end up taking over the custom skulltula message as well then we can just add the auto-dismissal here, or in a followup PR shortly afterwards. I think a bit more discussion on the scope of this needs to be had before I can say for sure.

m4xw committed:
git subrepo push soh

subrepo:
  subdir:   "soh"
  merged:   "75ccbade8"
upstream:
  origin:   "https://github.com/HarbourMasters/soh.git"
  branch:   "master"
  commit:   "75ccbade8"
git-subrepo:
  version:  "0.4.1"
  origin:   "???"
  commit:   "???"

leggettc18 committed:
Merge branch 'develop' of https://github.com/HarbourMasters/Shipwright into rando-custom-messages

leggettc18 committed:
Adds text for Bottle with Blue Fire as a demo.

leggettc18 commented:
![image](https://user-images.githubusercontent.com/21345721/179325214-592bf8e3-c906-4303-a472-34675d09bc66.png)

Here's what it looks like with `GI_BOTTLE_WITH_BLUE_FIRE` after my most recent commits. Check out the flow and let me know if there's anything that could be improved. I've actually got a couple of ideas already as far as better encapsulating everything, but I'll work on that this weekend.

leggettc18 commented:
The basic flow, in case it isn't immediately obvious.

1. Player gets an item (gets `player->getItemId` set to something).
2. During the execution path to give the item, the `getItemId` is looked up in the `sGetItemTable` array in z_player.c
3. The struct contained at that array index contains, among other things, that item's textId. For all custom items so far, this is 0xF8, since that is an unused textId and the definition for GetItemEntry type only allows it to be a `u8` so I don't have many options.
4. In `z_message_PAL.c`, this textId is intercepted (but only when rando is enabled) and instead my function `Randomizer_GetCustomGetItemMessage` which takes the getItemId (obtained from the player struct), the text buffer and the size, and puts the custom message in the text buffer.
5. Through a couple of other function calls, that information is eventually passed to `randomizer_custom_messages.cpp`'s `Randomizer::GetCustomGetItemMessage`, which does a switch case based on the Item ID, and then the language, to return a string composed by concatenating string literals and some helper functions for control codes.

leggettc18 committed:
Refactors CustomMessages to not be rando specific.

Pulls the Custom Message related code out into it's own class, which has an initialization phase where other enhancements / future features can create messages during an initialization phase to be stored and retrieved later. Along with this refactoring, the 4 bottle messages from the previous rando-specific system are now created and stored during intialization and retrieved by their getItemId.

Now that it isn't rando specific, the goal is to move anything text changes that are hard-coded into z_message_PAL.c and refactor it so that future text additions/overrides can be done without modifying that file.

leggettc18 commented:
I've refactored a lot of the CustomMessage code to not be rando-specific, and generate messages in advance during initialization rather than on the fly. Of course when it comes to things like Skulltula token auto-dismissal of textboxes and skullsanity there may end up needing to be on-the-fly transformations of certain strings during retrieval, but I think that's probably better than generating and formatting the entire string on the fly. Not really sure. Maybe I should also consider moving this PR to develop instead of rando-next, but I'm not sure. Stop me now if this is an unwanted direction.

leggettc18 committed:
Does some cleanup of unecessary code and cases of no string provided.

leggettc18 committed:
Implements function to return custom message if one exists.

`CustomMessage_RetrieveIfExists` is placed at the very beginning of the z_message_PAL.c `Message_OpenText` if statement after any special changes to textIds are made. This function will either return either true or false and if true it will populate the necessary values of font and msgCtx to display said message. It's placement at the beginning also allows for overriding existing textIds to have new information, which will come in handy later.

leggettc18 committed:
Simplifies the function signature of `CustomMessage_RetrieveIfExists`.

leggettc18 committed:
more cleanup

leggettc18 committed:
Implements adding message tables and retrieving by an ID.

Basically, some external code can choos a unique id and create a message table for itself. The idea is that modders can hook into this as well so they can get their own message table and don't have to worry about conflicting with the base game's textIDs. I have also implemented this setup for Randomizer so that the pre-filled bottles (which didn't exist in the original outside of Ruto's letter) have unique text as compared to the text for the item they are filled with.

leggettc18 committed:
Implements the custom text for gInjectSkulltulaCount.

Also auto-dissmisses the textbox when gSkulltulaFreeze is on. When gInjectSkulltulaCount is off, text is not auto-dismissed just like before, since that is base game text still.

leggettc18 commented:
This has taken shape to be something that is useful for rando, but not rando specific, so for now I will be opening a new PR on Develop. Also since we're having issues with this branch being behind develop anyway. Some of my earlier comments aren't that relevant anymore so I'd like to start fresh now that this has taken a bit more shape.
